lake/one_shot_dataset.py

```diff
@@ -0,0 +1,251 @@
+
+import os
+
+from PIL import Image
+
+import torch
+from torch.utils.data import Dataset
+
+import torchvision
+from torchvision.datasets.utils import check_integrity
+
+import numpy as np
+
+import imageio
+from scipy import ndimage
+
+from pathlib import Path
+from googleapiclient.discovery import build
+import io
+from googleapiclient.http import MediaIoBaseDownload
+import shutil
+import zipfile
+
+class ImageTransformation:
+  """Transform images by resizing and centring"""
+
+  def __init__(self, centre=True, invert=True, resize_factor=1.0):
+    self.centre = centre
+    self.resize_factor = resize_factor
+
+  def __call__(self, x):
+    # Resize
+    if self.resize_factor != 1.0:
+      height = int(self.resize_factor * x.shape[1])
+      width = int(self.resize_factor * x.shape[2])
+
+      x = torchvision.transforms.ToPILImage()(x)
+      x = torchvision.transforms.functional.resize(x, size=[height, width])
+
+      x = torchvision.transforms.functional.to_tensor(x)
+
+
+    # Centre the image
+    if self.centre:
+      # NCHW => NHWC
+      x = x.permute(1, 2, 0)
+
+      # Compute centre
+      centre = np.array([int(x.shape[0]) * 0.5, int(x.shape[1]) * 0.5])
+
+      # Compute centre of mass
+      centre_of_mass = ndimage.measurements.center_of_mass(x.numpy())
+      centre_of_mass = np.array(centre_of_mass[:-1])
+
+      # Compute translation
+      translation = (centre - centre_of_mass).tolist()
+      translation.reverse()
+
+      # Apply transformation
+      # NHWC => NCHW
+      x = x.permute(2, 0, 1)
+      x = torchvision.transforms.ToPILImage()(x)
+      x = torchvision.transforms.functional.affine(x, 0, translation, scale=1.0, shear=0, resample=Image.BILINEAR)
+
+      # Convert back to tensor
+      x = torchvision.transforms.functional.to_tensor(x)
+
+    return x
+
+
+class OneShotDataset(Dataset):
```

Comment:
Is this class used?

Reply:
I will delete this class, it was an idea that I had that I can implement later (this code doesn't really do anything). 

The idea is that there is one dataset class for all datasets. Will be a bit complicated to implement but if we are thinking of introducing more datasets later on it will make the code for oneshot_cls much more concise.